### PR TITLE
Introduce `WP_CLI\Utils\basename()` for locale-independent basename()

### LINKED ISSUE #3818: basename() doesn't deal well with UTF-8.
As mentioned in core trac [#33227](https://core.trac.wordpress.org/ticket/33227#comment:12)) `basename()` and `PATHINFO_BASENAME` are PHP locale dependent and fail with UTF-8 file names unless the locale is UTF-8.

A simple fix is just to replace `basename()` with the locale-independent `wp_basename()`.

### PR DESCRIPTION
**Pull request by danielbachhuber**
Fixes #3818

### CODE REVIEW

File: php/utils.php [hunk 1]
```diff
@@ -823,3 +823,16 @@ function parse_str_to_argv( $arguments ) {
 	}, $argv );
 	return $argv;
 }
+
+/**
+ * Locale-independent version of basename()
+ *
+ * @access public
```
Comment (schlessera) [inline php/utils.php]: WPCS: Functions should only have an `@access` tag if they are private.

File: php/utils.php [hunk 2]
```diff
@@ -823,3 +823,16 @@ function parse_str_to_argv( $arguments ) {
 	}, $argv );
 	return $argv;
 }
+
+/**
+ * Locale-independent version of basename()
```
Comment (schlessera) [inline php/utils.php]: WPCS: Missing period.

File: php/utils.php [hunk 3]
```diff
@@ -823,3 +823,16 @@ function parse_str_to_argv( $arguments ) {
...
+ *
+ * @access public
+ *
+ * @param string $path
```
Comment (schlessera) [inline php/utils.php]: Not sure what's your stance on this. Shouldn't we make sure all doc blocks contain complete information?
In this case, short summaries, with an `Optional.` and the default value for `$suffix` are needed.

File: php/commands/media.php [hunk 4]
```diff
@@ -216,7 +216,7 @@ function import( $args, $assoc_args = array() ) {
 
 			$file_array = array(
 				'tmp_name' => $tempfile,
-				'name' => basename( $file )
+				'name' => Utils\basename( $file )
```
Comment (schlessera) [inline php/commands/media.php]: WPCS: trailing comma recommended

File: php/utils.php (hunk 1 above)
Comment (danielbachhuber) [inline php/utils.php]: `@access public` is how some utils end up here: http://wp-cli.org/docs/internal-api/

**Comment by danielbachhuber** [comment]
@schlessera I'd suggest we revisit WPCS changes at a later date, possibly in conjunction with https://github.com/wp-cli/scaffold-package-command/issues/70. It's not something I've strictly enforced to date.

**Pull request merged**